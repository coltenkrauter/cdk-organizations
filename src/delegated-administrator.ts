import { RemovalPolicy } from "aws-cdk-lib";
import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from "aws-cdk-lib/custom-resources";
import { Construct } from "constructs";
import { IAccount } from "./account";

export interface DelegatedAdministratorProps {
  /**
   * The member account in the organization to register as a delegated administrator.
   */
  readonly account: IAccount;
  /**
   * The service principal of the AWS service for which you want to make the member account a delegated administrator.
   */
  readonly servicePrincipal: string;
  /**
   * The region to delegate the administrator in.
   */
  readonly region?: string;
  /**
   * If set to RemovalPolicy.RETAIN, the delegation will not be removed.
   *
   * @default RemovalPolicy.DESTROY
   */
  readonly removalPolicy?: RemovalPolicy;
}

/**
 * Enables the specified member account to administer the Organizations features of the specified AWS service. It grants read-only access to AWS Organizations service data. The account still requires IAM permissions to access and administer the AWS service.
 *
 * You can run this action only for AWS services that support this feature. For a current list of services that support it, see the column Supports Delegated Administrator in the table at AWS Services that you can use with AWS Organizations in the [AWS Organizations User Guide](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_integrate_services_list.html).
 *
 * @see https://docs.aws.amazon.com/accounts/latest/reference/using-orgs-delegated-admin.html
 */
export class DelegatedAdministrator extends Construct {
  public constructor(scope: Construct, id: string, props: DelegatedAdministratorProps) {
    super(scope, id);

    const { account, servicePrincipal, region } = props;

    new AwsCustomResource(this, "DelegatedAdministratorCustomResource", {
      resourceType: "Custom::Organizations_DelegatedAdministrator",
      onCreate: {
        service: "Organizations",
        action: "registerDelegatedAdministrator", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#registerDelegatedAdministrator-property
        region: region ?? "us-east-1",
        physicalResourceId: PhysicalResourceId.of(`${account.accountId}:${servicePrincipal}`),
        parameters: {
          AccountId: account.accountId,
          ServicePrincipal: servicePrincipal,
        },
        ignoreErrorCodesMatching: "AccountAlreadyRegisteredException", // https://docs.aws.amazon.com/organizations/latest/APIReference/API_RegisterDelegatedAdministrator.html#API_RegisterDelegatedAdministrator_Errors
      },
      ...(props.removalPolicy === RemovalPolicy.RETAIN
        ? {}
        : {
            onDelete: {
              service: "Organizations",
              action: "deregisterDelegatedAdministrator", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#deregisterDelegatedAdministrator-property
              region: region ?? "us-east-1",
              parameters: {
                AccountId: account.accountId,
                ServicePrincipal: servicePrincipal,
              },
            },
          }),
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}
