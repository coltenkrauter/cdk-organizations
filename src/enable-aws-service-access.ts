import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from "aws-cdk-lib/custom-resources";
import { Construct } from "constructs";

export interface EnableAwsServiceAccessProps {
  /**
   * The service principal name of the AWS service for which you want to enable integration with your organization. This is typically in the form of a URL, such as service-abbreviation.amazonaws.com.
   */
  readonly servicePrincipal: string;
}

/**
 * Enables the integration of an AWS service (the service that is specified by ServicePrincipal) with AWS Organizations. When you enable integration, you allow the specified service to create a service-linked role in all the accounts in your organization. This allows the service to perform operations on your behalf in your organization and its accounts.
 *
 * <strong>This operation can be called only from the organization's management account and only if the organization has enabled all features.</strong>
 *
 * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_integrate_services.html#orgs_trusted_access_perms
 */
export class EnableAwsServiceAccess extends Construct {
  public constructor(scope: Construct, id: string, props: EnableAwsServiceAccessProps) {
    super(scope, id);

    const { servicePrincipal } = props;
    const organizationsRegion = process.env.CDK_AWS_PARTITION === "aws-cn" ? "cn-northwest-1" : "us-east-1";

    new AwsCustomResource(this, "EnableAwsServiceAccessCustomResource", {
      resourceType: "Custom::Organizations_EnableAwsServiceAccess",
      onCreate: {
        service: "Organizations",
        action: "enableAWSServiceAccess", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#enableAWSServiceAccess-property
        region: organizationsRegion,
        physicalResourceId: PhysicalResourceId.of(`${servicePrincipal}`),
        parameters: {
          ServicePrincipal: servicePrincipal,
        },
      },
      onDelete: {
        service: "Organizations",
        action: "disableAWSServiceAccess", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#disableAWSServiceAccess-property
        region: organizationsRegion,
        parameters: {
          ServicePrincipal: servicePrincipal,
        },
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}
