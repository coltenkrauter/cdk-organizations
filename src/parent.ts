import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from "aws-cdk-lib/custom-resources";
import { Construct, IConstruct } from "constructs";
import { IResource } from "./resource";

export interface IParent extends IConstruct, IResource {}

export interface IChild extends IConstruct, IResource {}

export interface ParentProps {
  readonly child: IChild;
}

export interface ParentBaseProps {
  readonly childId: string;
}

export abstract class ParentBase extends Construct implements IParent {
  public readonly parentId: string;

  protected constructor(scope: Construct, id: string, props: ParentBaseProps) {
    super(scope, id);

    const { childId } = props;
    const organizationsRegion = process.env.CDK_AWS_PARTITION === "aws-cn" ? "cn-northwest-1" : "us-east-1";

    const parent = new AwsCustomResource(this, "ListParentsCustomResource", {
      onCreate: {
        service: "Organizations",
        action: "listParents", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#listParents-property
        region: organizationsRegion,
        physicalResourceId: PhysicalResourceId.fromResponse("Parents.0.Id"),
        parameters: {
          ChildId: childId,
        },
      },
      onUpdate: {
        service: "Organizations",
        action: "listParents", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#listParents-property
        region: organizationsRegion,
        physicalResourceId: PhysicalResourceId.fromResponse("Parents.0.Id"),
        parameters: {
          ChildId: childId,
        },
      },
      onDelete: {
        service: "Organizations",
        action: "listParents", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#listParents-property
        region: organizationsRegion,
        parameters: {
          ChildId: childId,
        },
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });

    this.parentId = parent.getResponseField("Parents.0.Id");
  }

  public identifier(): string {
    return this.parentId;
  }
}

export class Parent extends ParentBase {
  public static fromChildId(scope: Construct, id: string, childId: string): IParent {
    class Import extends ParentBase {
      public constructor() {
        super(scope, id, { childId: childId });
      }
    }

    return new Import();
  }

  public constructor(scope: Construct, id: string, props: ParentProps) {
    const { child } = props;

    super(scope, id, { childId: child.identifier() });

    this.node.addDependency(child);
  }
}
