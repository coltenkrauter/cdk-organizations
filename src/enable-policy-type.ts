import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from "aws-cdk-lib/custom-resources";
import { Construct } from "constructs";
import { Root } from "./organization";
import { PolicyType } from "./policy";

export interface EnablePolicyTypeProps {
  readonly root: Root;
  readonly policyType: PolicyType;
}

/**
 * Enables and disables Enables a policy type in a root. After you enable a policy type in a root, you can attach policies of that type to the root, any organizational unit (OU), or account in that root.
 *
 * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_enable-disable.html
 */
export class EnablePolicyType extends Construct {
  public constructor(scope: Construct, id: string, props: EnablePolicyTypeProps) {
    super(scope, id);

    const { root, policyType } = props;
    const organizationsRegion = process.env.CDK_AWS_PARTITION === "aws-cn" ? "cn-northwest-1" : "us-east-1";

    new AwsCustomResource(this, "EnablePolicyTypeCustomResource", {
      resourceType: "Custom::Organizations_EnablePolicyType",
      onCreate: {
        service: "Organizations",
        action: "enablePolicyType", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#enablePolicyType-property
        region: organizationsRegion,
        physicalResourceId: PhysicalResourceId.of(`${root.rootId}:${policyType}`),
        parameters: {
          RootId: root.rootId,
          PolicyType: policyType,
        },
        ignoreErrorCodesMatching: "PolicyTypeAlreadyEnabledException", // https://docs.aws.amazon.com/organizations/latest/APIReference/API_EnablePolicyType.html#API_EnablePolicyType_Errors
      },
      onDelete: {
        service: "Organizations",
        action: "disablePolicyType", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#disablePolicyType-property
        region: organizationsRegion,
        parameters: {
          RootId: root.rootId,
          PolicyType: policyType,
        },
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}
