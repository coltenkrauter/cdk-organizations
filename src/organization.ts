import { Aspects, CustomResource, Names, Stack, TagManager, TagType, Token } from "aws-cdk-lib";
import * as aws_iam from "aws-cdk-lib/aws-iam";
import * as custom_resources from "aws-cdk-lib/custom-resources";
import { Construct, IConstruct } from "constructs";
import { pascalCase } from "pascal-case";
import { DependencyChain } from "./dependency-chain";
import { EnableAwsServiceAccess } from "./enable-aws-service-access";
import { EnablePolicyType } from "./enable-policy-type";
import { OrganizationProvider } from "./organization-provider";
import { IParent } from "./parent";
import { IPolicy, PolicyType } from "./policy";
import { IPolicyAttachmentTarget, PolicyAttachment } from "./policy-attachment";
import { ITaggableResource, TagResource } from "./tag-resource";

/**
 * Specifies the feature set supported by the new organization. Each feature set supports different levels of functionality.
 *
 * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_getting-started_concepts.html#feature-set
 */
export enum FeatureSet {
  /**
   * All member accounts have their bills consolidated to and paid by the management account. For more information, see [Consolidated billing](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_getting-started_concepts.html#feature-set-cb-only) in the AWS Organizations User Guide. The consolidated billing feature subset isnâ€™t available for organizations in the AWS GovCloud (US) Region.
   */
  CONSOLIDATED_BILLING = "CONSOLIDATED_BILLING",
  /**
   * In addition to all the features supported by the consolidated billing feature set, the management account can also apply any policy type to any member account in the organization. For more information, see [All features](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_getting-started_concepts.html#feature-set-all) in the AWS Organizations User Guide.
   */
  ALL = "ALL",
}

export interface OrganizationProps {
  /**
   * Enabling features in your organization.
   *
   * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_org_support-all-features.html
   *
   * @default ALL
   */
  readonly featureSet?: FeatureSet;
}

/**
 * Creates an organization to consolidate your AWS accounts so that you can administer them as a single unit. An organization has one management account along with zero or more member accounts. You can organize the accounts in a hierarchical, tree-like structure with a root at the top and organizational units nested under the root. Each account can be directly in the root, or placed in one of the OUs in the hierarchy. An organization has the functionality that is determined by the feature set that you enable.
 *
 * <strong>The account whose user is calling the CreateOrganization operation automatically becomes the management account of the new organization.</strong>
 *
 * <strong>For deletion of an organization you must previously remove all the member accounts, OUs, and policies from the organization!</strong>
 *
 * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_org_create.html#create-org
 */
export interface IOrganization extends IConstruct {
  /**
   * The unique identifier (ID) of an organization. The regex pattern for an organization ID string requires "o-" followed by from 10 to 32 lowercase letters or digits.
   */
  readonly organizationId: string;
  /**
   * The Amazon Resource Name (ARN) of an organization.
   */
  readonly organizationArn: string;
  /**
   * Specifies the functionality that currently is available to the organization. If set to "ALL", then all features are enabled and policies can be applied to accounts in the organization. If set to "CONSOLIDATED_BILLING", then only consolidated billing functionality is available.
   */
  readonly featureSet: FeatureSet;
  /**
   * The Amazon Resource Name (ARN) of the account that is designated as the management account for the organization.
   */
  readonly managementAccountArn: string;
  /**
   * The unique identifier (ID) of the management account of an organization.
   */
  readonly managementAccountId: string;
  /**
   * The email address that is associated with the AWS account that is designated as the management account for the organization.
   */
  readonly managementAccountEmail: string;
  /**
   * The principal that represents this AWS Organization
   */
  readonly principal: aws_iam.IPrincipal;
}

export class Organization extends Construct implements IOrganization {
  /**
   * Describe the organization that the current account belongs to.
   *
   * @see https://docs.aws.amazon.com/organizations/latest/APIReference/API_DescribeOrganization.html
   */
  public static of(scope: Construct, id: string): IOrganization {
    class Import extends Construct implements IOrganization {
      readonly featureSet: FeatureSet;
      readonly managementAccountArn: string;
      readonly managementAccountEmail: string;
      readonly managementAccountId: string;
      readonly organizationArn: string;
      readonly organizationId: string;
      readonly principal: aws_iam.IPrincipal;

      public constructor() {
        super(scope, id);

        const organizationsRegion = process.env.CDK_AWS_PARTITION === "aws-cn" ? "cn-northwest-1" : "us-east-1";

        const resource = new custom_resources.AwsCustomResource(scope, "CustomResource", {
          resourceType: "Custom::Organizations_ImportOrganization",
          onCreate: {
            service: "Organizations",
            action: "describeOrganization", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#describeOrganization-property
            region: organizationsRegion,
            parameters: {},
            physicalResourceId: custom_resources.PhysicalResourceId.fromResponse("Organization.Id"),
          },
          onUpdate: {
            service: "Organizations",
            action: "describeOrganization", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#describeOrganization-property
            region: organizationsRegion,
            parameters: {},
            physicalResourceId: custom_resources.PhysicalResourceId.fromResponse("Organization.Id"),
          },
          installLatestAwsSdk: false,
          policy: custom_resources.AwsCustomResourcePolicy.fromSdkCalls({
            resources: custom_resources.AwsCustomResourcePolicy.ANY_RESOURCE,
          }),
        });

        this.featureSet = resource.getResponseField("Organization.FeatureSet") as FeatureSet;
        this.managementAccountArn = resource.getResponseField("Organization.MasterAccountArn");
        this.managementAccountEmail = resource.getResponseField("Organization.MasterAccountEmail");
        this.managementAccountId = resource.getResponseField("Organization.MasterAccountId");
        this.organizationArn = resource.getResponseField("Organization.Arn");
        this.organizationId = resource.getResponseField("Organization.Id");
        this.principal = new aws_iam.OrganizationPrincipal(this.organizationId);
      }
    }

    return new Import();
  }

  public readonly organizationId: string;
  public readonly organizationArn: string;
  public readonly featureSet: FeatureSet;
  public readonly managementAccountArn: string;
  public readonly managementAccountId: string;
  public readonly managementAccountEmail: string;
  readonly principal: aws_iam.IPrincipal;
  /**
   * The root of the current organization, which is automatically created.
   */
  readonly root: Root;

  private readonly resource: CustomResource;

  public constructor(scope: Construct, id: string, props: OrganizationProps = {}) {
    super(scope, id);

    const featureSet = props.featureSet || FeatureSet.ALL;

    const organizationProvider = OrganizationProvider.getOrCreate(this);
    this.resource = new CustomResource(this, "Organization", {
      serviceToken: organizationProvider.provider.serviceToken,
      resourceType: "Custom::Organizations_Organization",
      properties: {
        FeatureSet: featureSet,
      },
    });

    this.organizationId = this.resource.getAtt("Id").toString();
    this.organizationArn = this.resource.getAtt("Arn").toString();
    this.featureSet = this.resource.getAtt("FeatureSet").toString() as FeatureSet;
    this.managementAccountArn = this.resource.getAtt("MasterAccountArn").toString();
    this.managementAccountId = this.resource.getAtt("MasterAccountId").toString();
    this.managementAccountEmail = this.resource.getAtt("MasterAccountEmail").toString();

    // Use a dummy organization ID for validation during synthesis
    // The actual ID will be used at deployment time
    this.principal = Token.isUnresolved(this.organizationId)
      ? new aws_iam.OrganizationPrincipal("o-0000000000")
      : new aws_iam.OrganizationPrincipal(this.organizationId);

    this.root = new Root(this, "Root");
    this.root.node.addDependency(this.resource);
  }

  /**
   * Enables trusted access for a supported AWS service (trusted service), which performs tasks in your organization and its accounts on your behalf.
   * @param servicePrincipal The supported AWS service that you specify
   *
   * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_integrate_services_list.html
   */
  public enableAwsServiceAccess(servicePrincipal: string) {
    const enableAwsServiceAccess = new EnableAwsServiceAccess(this, `Enable${pascalCase(servicePrincipal)}`, {
      servicePrincipal: servicePrincipal,
    });
    enableAwsServiceAccess.node.addDependency(this.resource);
  }

  /**
   * Enables policy types in the following two broad categories: Authorization policies and Management policies.
   * @param policyType: the type of the policy that you specify
   *
   * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies.html#orgs-policy-types
   */
  public enablePolicyType(policyType: PolicyType) {
    this.root.enablePolicyType(policyType);
  }

  /**
   * Attach a policy. Before you can attach the policy, you must enable that policy type for use. You can use policies when you have all features enabled.
   *
   * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies.html
   */
  public attachPolicy(policy: IPolicy) {
    this.root.attachPolicy(policy);
  }
}

/**
 * The parent container for all the accounts for your organization. If you apply a policy to the root, it applies to all organizational units (OUs) and accounts in the organization.
 * <strong>Currently, you can have only one root. AWS Organizations automatically creates it for you when you create an organization.</strong>
 * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_getting-started_concepts.html
 */
export class Root extends Construct implements IParent, IPolicyAttachmentTarget, ITaggableResource {
  /**
   * The unique identifier (ID) for the root. The regex pattern for a root ID string requires "r-" followed by from 4 to 32 lowercase letters or digits.
   */
  public readonly rootId: string;

  protected readonly resource: custom_resources.AwsCustomResource;

  private readonly scope: Construct;

  readonly tags = new TagManager(TagType.KEY_VALUE, "Custom::Organizations_Root");

  public constructor(scope: Construct, id: string) {
    super(scope, id);
    this.scope = scope;

    const organizationsRegion = process.env.CDK_AWS_PARTITION === "aws-cn" ? "cn-northwest-1" : "us-east-1";

    this.resource = new custom_resources.AwsCustomResource(this, "RootCustomResource", {
      resourceType: "Custom::Organizations_Root",
      onCreate: {
        service: "Organizations",
        action: "listRoots", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#listRoots-property
        region: organizationsRegion,
        physicalResourceId: custom_resources.PhysicalResourceId.fromResponse("Roots.0.Id"),
      },
      onUpdate: {
        service: "Organizations",
        action: "listRoots", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#listRoots-property
        region: organizationsRegion,
        physicalResourceId: custom_resources.PhysicalResourceId.fromResponse("Roots.0.Id"),
      },
      onDelete: {
        service: "Organizations",
        action: "listRoots", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#listRoots-property
        region: organizationsRegion,
      },
      installLatestAwsSdk: false,
      policy: custom_resources.AwsCustomResourcePolicy.fromSdkCalls({
        resources: custom_resources.AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });

    this.rootId = this.resource.getResponseField("Roots.0.Id"); // Returns first root id. It seems AWS Organizations doesn't contain multiple roots.

    const stack = Stack.of(this);
    Aspects.of(stack).add(new DependencyChain()); // sequentially chain organization resources which can't be deployed in parallel

    const tagResource = new TagResource(this, "Tags", { resourceId: this.rootId, tags: this.tags.renderedTags });
    tagResource.node.addDependency(this.resource);
  }

  public identifier(): string {
    return this.rootId;
  }

  /**
   * Attach a policy. Before you can attach the policy, you must enable that policy type for use. You can use policies when you have all features enabled.
   *
   * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies.html
   */
  public attachPolicy(policy: IPolicy) {
    const policyAttachment = new PolicyAttachment(
      this.scope,
      `PolicyAttachment-${Names.nodeUniqueId(this.node)}-${Names.nodeUniqueId(policy.node)}`,
      {
        target: this,
        policy: policy,
      }
    );
    policyAttachment.node.addDependency(this.resource, policy);
  }

  /**
   * Enables and disables Enables a policy type. After you enable a policy type in a root, you can attach policies of that type to the root, any organizational unit (OU), or account in that root.
   *
   * @see https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_enable-disable.html
   */
  public enablePolicyType(policyType: PolicyType) {
    const enablePolicyType = new EnablePolicyType(this.scope, `Enable${pascalCase(policyType)}`, {
      root: this,
      policyType: policyType,
    });
    enablePolicyType.node.addDependency(this.resource);
  }
}
