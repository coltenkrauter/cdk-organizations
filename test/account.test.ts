import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Account, Organization } from "../src";

describe("Account", () => {
  it("Should match snapshot", () => {
    // Given
    const stack = new Stack(undefined, undefined, { env: { account: "123456789012", region: "us-east-1" } });
    const organization = new Organization(stack, "Organization", {});

    // When
    new Account(stack, "Account", {
      email: "info@pepperize.com",
      accountName: "test",
      parent: organization.root,
    });

    // Then
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });

  it("Should have delegated administrator", () => {
    // Given
    const stack = new Stack();
    const organization = new Organization(stack, "Organization", {});
    const account = new Account(stack, "Account", {
      email: "info@pepperize.com",
      accountName: "test",
      parent: organization.root,
    });

    // When
    account.delegateAdministrator("service-abbreviation.amazonaws.com");

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs("Custom::Organizations_DelegatedAdministrator", 1);
  });

  it("Should have delegated region administrator", () => {
    // Given
    const stack = new Stack();
    const organization = new Organization(stack, "Organization", {});
    const account = new Account(stack, "Account", {
      email: "info@pepperize.com",
      accountName: "test",
      parent: organization.root,
    });

    // When
    account.delegateAdministrator("service-abbreviation.amazonaws.com", "eu-west-1");

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs("Custom::Organizations_DelegatedAdministrator", 1);
  });
});
