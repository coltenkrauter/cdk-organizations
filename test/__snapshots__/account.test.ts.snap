// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Should match snapshot 1`] = `
Object {
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AccountCreateAccount833709C2": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccountName": "test",
        "Email": "info@pepperize.com",
        "IamUserAccessToBilling": "ALLOW",
        "ImportOnDuplicate": "true",
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "OrganizationRootRootCustomResourceBB74F060",
            "Roots.0.Id",
          ],
        },
        "RemovalPolicy": "retain",
        "RoleName": "OrganizationAccountAccessRole",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "cdkorganizationsAccountProviderNestedStackcdkorganizationsAccountProviderNestedStackResourceA1C2E3D5",
            "Outputs.cdkorganizationsAccountProviderframeworkonEvent4241E2B3Arn",
          ],
        },
      },
      "Type": "Custom::Organizations_Account",
      "UpdateReplacePolicy": "Delete",
    },
    "AccountTagsTagResourceB6D57C22": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "AccountCreateAccount833709C2",
      ],
      "Properties": Object {
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "AccountCreateAccount833709C2",
            "AccountId",
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "cdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResourceA7B8DF96",
            "Outputs.cdkorganizationsTagResourceProviderframeworkonEventDD009DFBArn",
          ],
        },
      },
      "Type": "Custom::Organizations_TagResource",
      "UpdateReplacePolicy": "Delete",
    },
    "Organization06E16095": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "FeatureSet": "ALL",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "cdkorganizationsOrganizationProviderNestedStackcdkorganizationsOrganizationProviderNestedStackResourceE0751832",
            "Outputs.cdkorganizationsOrganizationProviderframeworkonEvent268B5E2CArn",
          ],
        },
      },
      "Type": "Custom::Organizations_Organization",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationRootRootCustomResourceBB74F060": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "Organization06E16095",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Roots.0.Id\\"}}",
        "Delete": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\"}",
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Roots.0.Id\\"}}",
      },
      "Type": "Custom::Organizations_Root",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E": Object {
      "DependsOn": Array [
        "Organization06E16095",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "organizations:ListRoots",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OrganizationRootTagsTagResourceCBEA7B2F": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "Organization06E16095",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
      ],
      "Properties": Object {
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "OrganizationRootRootCustomResourceBB74F060",
            "Roots.0.Id",
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "cdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResourceA7B8DF96",
            "Outputs.cdkorganizationsTagResourceProviderframeworkonEventDD009DFBArn",
          ],
        },
      },
      "Type": "Custom::Organizations_TagResource",
      "UpdateReplacePolicy": "Delete",
    },
    "cdkorganizationsAccountProviderNestedStackcdkorganizationsAccountProviderNestedStackResourceA1C2E3D5": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.us-east-1.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/cdk-hnb659fds-assets-123456789012-us-east-1/7ae50027841ef2f97a51eef91fcb045c88387f981533e58c9de888f872989e2a.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "cdkorganizationsOrganizationProviderNestedStackcdkorganizationsOrganizationProviderNestedStackResourceE0751832": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.us-east-1.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/cdk-hnb659fds-assets-123456789012-us-east-1/36df93481d903a6d5f1b314e99a5886f19fad6eb694c60de67d8b6c33fd76fe4.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "cdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResourceA7B8DF96": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.us-east-1.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/cdk-hnb659fds-assets-123456789012-us-east-1/a8fbc353a713f63c9e1159c03781e99021638487ae87ad16ec30ac7a42a351e9.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
